x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/MetricFieldProducer.java

```diff
@@ -0,0 +1,237 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.rollup.v2;
+
+import org.elasticsearch.index.mapper.MappedFieldType;
+import org.elasticsearch.index.query.SearchExecutionContext;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+abstract class MetricFieldProducer {
+    private final String field;
+    private final List<Metric> metrics;
+    private boolean isEmpty = true;
+
+    MetricFieldProducer(String field, List<Metric> metrics) {
+        this.field = field;
+        this.metrics = metrics;
+    }
+
+    void reset() {
+        for (Metric metric : metrics) {
+            metric.reset();
+        }
+        isEmpty = true;
+    }
+
+    public String field() {
+        return field;
+    }
+
+    public List<Metric> metrics() {
+        return metrics;
+    }
+
+    public void collectMetric(Double value) {
+        for (MetricFieldProducer.Metric metric : metrics) {
+            metric.collect(value);
+        }
+        isEmpty = false;
+    }
+
+    public boolean isEmpty() {
+        return isEmpty;
+    }
+
+    public abstract Object value();
+
+    abstract static class Metric {
+        final String name;
+
+        protected Metric(String name) {
+            this.name = name;
+        }
+
+        abstract void collect(double number);
+
+        abstract Number get();
+
+        abstract void reset();
+    }
+
+    static class Max extends Metric {
+        private Double max;
```

Comment:
I see we are using double here to keep track of metrics. In the aggregation framework using double as the underlying common data type is a source of issues/pain. Should we consider using BigDecimal? Precision would benefit...

Reply:
You are right that double is potentially a source of drouble. However, all our aggs already use doubles everywhere. Also, `aggregate_metric_double` will store the result as a double